README.md
# perl-Gamstop

This module provides an interface to [GAMSTOP](https://www.gamstop.co.uk/) api.

GAMSTOP is an online self-exclusion service enabling consumers resident within
the UK to voluntarily exclude themselves from all British-licensed remote
gambling operators.

# NAME
    Webservice::GAMSTOP - GAMSTOP API Client Implementation

# SYNOPSIS
        use Webservice::GAMSTOP;
        my $instance = Webservice::GAMSTOP->new(
            api_url => '<gamstop_api_url>',
            api_key => '<gamstop_api_key>',
            # optional (defaults to 5 seconds)
            timeout => 10,
        );

        $instance->get_exclusion_for(
            first_name    => 'Harry',
            last_name     => 'Potter',
            email         => 'harry.potter@example.com',
            date_of_birth => '1970-01-01',
            postcode      => 'hp11aa',
        );

# DESCRIPTION
    This module implements a programmatic interface to
    [GAMSTOP](https://www.gamstop.co.uk/) api.

# PRE-REQUISITE
    Before you can use this module, you'll need to obtain your own "Unique
    API Key" from [GAMSTOP](https://www.gamstop.co.uk/).

# METHODS
  ## get_exclusion_for
    Given user details return Webservice::GAMSTOP::Response object
    Note: it dies if an error occur connecting to GAMSTOP API endpoint

   ### Required parameters
    *   first_name : First name of person, only 20 characters are
        significant

    *   last_name : Last name of person, only 20 characters are significant

    *   date_of_birth: Date of birth in ISO format (yyyy-mm-dd)

    *   email : Email address

    *   postcode : Postcode (spaces not significant)

   ### Optional parameters
    *   x_trace_id: A freeform field that is put into audit log that can be
        used by the caller to identify a request. This might be something to
        indicate the person being checked, a unique request ID, GUID, or a
        trace ID from a system such as zipkin.

   ### Return value
   *    A Webservice::GAMSTOP::Response object

# AUTHOR
    binary.com <cpan@binary.com>

# COPYRIGHT AND LICENSE
    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.


Changes
{{$NEXT}}

0.003     2018-04-10 05:38:13+00:00 UTC
  Remove name validations

0.002     2018-03-21 03:45:41+00:00 UTC
  Fix failing test

0.001     2018-03-08 02:18:35+00:00 UTC
  Initial release



Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.007.
use strict;
use warnings;

use 5.014000;

use ExtUtils::MakeMaker 6.48;

my %WriteMakefileArgs = (
  "ABSTRACT" => "GAMSTOP API Client Implementation",
  "AUTHOR" => "binary.com <BINARY\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.48"
  },
  "DISTNAME" => "Webservice-GAMSTOP",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.014000",
  "NAME" => "Webservice::GAMSTOP",
  "PREREQ_PM" => {
    "Mojo::UserAgent" => 0,
    "Moo" => 0,
    "indirect" => "0.37"
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::CheckDeps" => "0.010",
    "Test::Exception" => "0.43",
    "Test::More" => "0.98"
  },
  "VERSION" => "0.001",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Mojo::UserAgent" => 0,
  "Moo" => 0,
  "Test::CheckDeps" => "0.010",
  "Test::Exception" => "0.43",
  "Test::More" => "0.98",
  "indirect" => "0.37"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/Webservice/GAMSTOP/Response.pm
package Webservice::GAMSTOP::Response;

use strict;
use warnings;

# VERSION

=head1 NAME

Webservice::GAMSTOP::Response - Response object for get_exclusion_for sub

=head1 SYNOPSIS

    use Webservice::GAMSTOP;
    my $instance = Webservice::GAMSTOP->new(
        api_url => 'gamstop_api_url',
        api_key => 'gamstop_api_key',
        # optional (defaults to 5 seconds)
        timeout => 10,
    );

    my $response = $instance->get_exclusion_for(
        first_name    => 'Harry',
        last_name     => 'Potter',
        email         => 'harry.potter@example.com',
        date_of_birth => '1970-01-01',
        postcode      => 'hp11aa',
    );

    $response->is_excluded;
    $response->get_date;
    $response->get_unique_id;
    $response->get_exclusion;

=head1 DESCRIPTION

This object is returned as response for get_exclusion_for.

=cut

=head1 METHODS

Constructor

=head2 new

    use Webservice::GAMSTOP::Response;
    my $response = Webservice::GAMSTOP::Response->new(
        exclusion => '',
        date      => '',
        unique_id => '',
    );

=head3 Return value

A new Webservice::GAMSTOP::Response object

=cut

sub new {
    my ($class, %args) = @_;

    return bless {}, $class unless %args;

    return bless \%args, $class;
}

=head2 get_exclusion

Exclusion flag provided in response headers

GAMSTOP Response:

- When GAMSTOP returns a Y response the user is registered with the GAMSTOP
service with a valid current self-exclusion.

- When GAMSTOP returns an N response the user is not registered with the GAMSTOP
service.

- When GAMSTOP returns a P response the user was previously self-excluded using
the GAMSTOP service but their chosen minimum period of exclusion has lapsed
and they have requested to have their self-exclusion removed

=head3 Return value

returns GAMSTOP exclusion flag or undef if not present

=over 4

=back

=cut

sub get_exclusion {
    return shift->{exclusion};
}

=head2 get_unique_id

Unique id provided in response headers

=head3 Return value

=over 4

returns GAMSTOP unique id for request or undef if not present

=back

=cut

sub get_unique_id {
    return shift->{unique_id};
}

=head2 get_date

Date provided in response headers. Format: Tue, 27 Feb 2018 02:42:01 GMT

=head3 Return value

=over 4

returns GAMSTOP response date or undef if not present

=back

=cut

sub get_date {
    return shift->{date};
}

=head2 is_excluded

Indicates whether user is self excluded or not

=head3 Return value

=over 4

True if user is excluded on GAMSTOP i.e GAMSTOP return a Y response else false

=back

=cut

sub is_excluded {
    my $flag = shift->{exclusion};

    return ($flag // '') eq 'Y' ? 1 : 0;
}

1;
__END__

=head1 AUTHOR

binary.com <cpan@binary.com>

=head1 COPYRIGHT AND LICENSE

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Webservice/GAMSTOP.pm
package Webservice::GAMSTOP;
# ABSTRACT: GAMSTOP API Client Implementation

use strict;
use warnings;

use Moo;
use Email::Valid;
use Mojo::UserAgent;

use Webservice::GAMSTOP::Response;

our $VERSION = '0.002';

=head1 NAME

Webservice::GAMSTOP - GAMSTOP API Client Implementation

=head1 VERSION

version 0.001

=head1 SYNOPSIS

    use Webservice::GAMSTOP;
    my $instance = Webservice::GAMSTOP->new(
        api_url => 'gamstop_api_url',
        api_key => 'gamstop_api_key',
        # optional (defaults to 5 seconds)
        timeout => 10,
    );

    $instance->get_exclusion_for(
        first_name    => 'Harry',
        last_name     => 'Potter',
        email         => 'harry.potter@example.com',
        date_of_birth => '1970-01-01',
        postcode      => 'hp11aa',
    );

=head1 DESCRIPTION

This module implements a programmatic interface to
[GAMSTOP](https://www.gamstop.co.uk/) API.

=head1 PRE-REQUISITE

Before you can use this module, you'll need to obtain your
own "Unique API Key" from [GAMSTOP](https://www.gamstop.co.uk/).

=cut

=head1 ATTRIBUTES

L<Webservice::GAMSTOP> implements the following attributes

=head2 api_url

GAMSTOP API endpoint url (REQUIRED)

=cut

has api_url => (
    is       => 'ro',
    required => 1,
);

=head2 api_key

GAMSTOP API unique key for operator (REQUIRED)

=cut

has api_key => (
    is       => 'ro',
    required => 1,
);

=head2 timeout

Maximum amount of time in seconds establishing a
connection may take before getting canceled
(OPTIONAL - DEFAULT 5 seconds)

=cut

has timeout => (
    is      => 'ro',
    default => 5,
);

has _ua => (
    is => 'lazy',
);

sub _build__ua {
    return Mojo::UserAgent->new->connect_timeout(shift->timeout);
}

=head1 METHODS

=head2 get_exclusion_for

Given user details return L<Webservice::GAMSTOP::Response> object
Note: it dies if an error occur connecting to GAMSTOP API endpoint

=head3 Required parameters

=over 4

=item * first_name   : First name of person, only 20 characters are significant

=item * last_name    : Last name of person, only 20 characters are significant

=item * date_of_birth: Date of birth in ISO format (yyyy-mm-dd)

=item * email        : Email address

=item * postcode     : Postcode (spaces not significant)

=back

=head3  Optional parameters

=over 4

x_trace_id: A freeform field that is put into audit log that can be used
by the caller to identify a request. This might be something to indicate
the person being checked, a unique request ID, GUID, or a trace ID from
a system such as zipkin.

=back

=head3 Return value

=over 4

A L<Webservice::GAMSTOP::Response> object

=back

=cut

sub get_exclusion_for {
    my ($self, %args) = @_;

    die "Missing required parameter: first_name."    unless (exists $args{first_name});
    die "Missing required parameter: last_name."     unless (exists $args{last_name});
    die "Missing required parameter: date_of_birth." unless (exists $args{date_of_birth});
    die "Missing required parameter: email."         unless (exists $args{email});
    die "Missing required parameter: postcode."      unless (exists $args{postcode});

    # validate args
    die 'Invalid date of birth. Date of birth should be in ISO format (yyyy-mm-dd)'
        unless $args{'date_of_birth'} =~ /^(?:\d{4}-(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01]))$/;
    die 'Invalid email.' unless Email::Valid->address($args{email});
    die 'Invalid postcode.' unless $args{postcode} =~ /^[^+]{0,20}$/;

    # required parameters
    my $form_params = {
        firstName   => $args{first_name},
        lastName    => $args{last_name},
        dateOfBirth => $args{date_of_birth},
        email       => $args{email},
        postcode    => $args{postcode},
    };

    # optional parameters
    $form_params->{'X-Trace-Id'} = $args{x_trace_id} if $args{x_trace_id};

    my $ua = $self->_ua;
    $ua->on(
        start => sub {
            my ($useragent, $tx) = @_;
            $tx->req->headers->header('X-API-Key' => $self->api_key);
        });

    my $tx = $ua->post($self->api_url => form => $form_params);

    if (my $err = $tx->error) {
        die 'Error - ' . $err->{code} . ' response: ' . $err->{message} if $err->{code};
        die 'Error - Connection error: ' . $err->{message};
    }

    my $headers = $tx->success->headers;
    return Webservice::GAMSTOP::Response->new(
        exclusion => $headers->header('x-exclusion'),
        date      => $headers->header('date'),
        unique_id => $headers->header('x-unique-id'),
    );
}

1;
__END__

=head1 AUTHOR

binary.com <cpan@binary.com>

=head1 COPYRIGHT AND LICENSE

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
